test/e2e_node/container_lifecycle_test.go

```diff
@@ -2875,22 +3606,148 @@ var _ = SIGDescribe(nodefeature.SidecarContainers, "Containers Lifecycle", func(
+			regular1 := "regular-1"
+
+			updatedImage := busyboxImage
+
+			var originalPodSpec *v1.Pod
+
+			ginkgo.BeforeEach(func(ctx context.Context) {
+				originalPodSpec = &v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "restartable-init-container-failed-startup-imgupdate",
+					},
+					Spec: v1.PodSpec{
+						RestartPolicy: v1.RestartPolicyNever,
+						InitContainers: []v1.Container{
+							{
+								Name:  restartableInit1,
+								Image: agnhostImage,
+								Command: ExecCommand(restartableInit1, execCommand{
+									Delay:    600,
+									ExitCode: 0,
+								}),
+								RestartPolicy: &containerRestartPolicyAlways,
+							},
+							{
+								Name:  restartableInit2,
+								Image: agnhostImage,
+								Command: ExecCommand(restartableInit2, execCommand{
+									Delay:    600,
+									ExitCode: 0,
+								}),
+								StartupProbe: &v1.Probe{
+									InitialDelaySeconds: 20,
+									FailureThreshold:    1,
+									ProbeHandler: v1.ProbeHandler{
+										Exec: &v1.ExecAction{
+											Command: []string{
+												"sh",
+												"-c",
+												"exit 1",
+											},
+										},
+									},
+								},
+								RestartPolicy: &containerRestartPolicyAlways,
+							},
+						},
+						Containers: []v1.Container{
+							{
+								Name:  regular1,
+								Image: agnhostImage,
+								Command: ExecCommand(regular1, execCommand{
+									Delay:    1,
+									ExitCode: 0,
+								}),
+							},
+						},
+					},
+				}
+			})
+
+			restartableInitContainerFailedStartupImageUpdateTest := func(ctx context.Context) {
+				preparePod(originalPodSpec)
+
+				client := e2epod.NewPodClient(f)
+				pod := client.Create(ctx, originalPodSpec)
+
+				ginkgo.By("Waiting for the restartable init container to restart", func() {
+					err := WaitForPodInitContainerRestartCount(ctx, f.ClientSet, pod.Namespace, pod.Name, 1, 1, 2*time.Minute)
+					framework.ExpectNoError(err)
+				})
+
+				ginkgo.By("Changing the image of the initializing restartable init container", func() {
+					client.Update(ctx, pod.Name, func(pod *v1.Pod) {
+						pod.Spec.InitContainers[1].Image = updatedImage
+					})
+				})
+
+				ginkgo.By("verifying that the image changed", func() {
+					err := e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "pod should be pending",
+						2*time.Minute, func(pod *v1.Pod) (bool, error) {
+							return pod.Status.Phase == v1.PodPending, nil
+						})
+					framework.ExpectNoError(err)
+
+					err = WaitForPodInitContainerRestartCount(ctx, f.ClientSet, pod.Namespace, pod.Name, 1, 2, 2*time.Minute)
+					framework.ExpectNoError(err)
+
+					err = e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "init container attempted to run with updated image",
+						time.Duration(30)*time.Second, func(pod *v1.Pod) (bool, error) {
+							containerStatus := pod.Status.InitContainerStatuses[1]
+							return containerStatus.Image == updatedImage, nil
```

Comment:
should this also check the restartCount? How do we know that it was actually restarted?

Reply:
It's checked in the previous line, but will add it here as well